### PR TITLE
drivers: spi: fix 16 bit spi dma transfers for the STM32 driver

### LINKED ISSUE #52563: spi_transceive with DMA for STM32 does not work for devices with 16 bit words.
**Describe the bug**
The STM32 SPI driver's DMA implementation does not work with devices that utilize 16 bit SPI words. This seems to be due to a bug in the STM32's SPI driver code. Currently, when calculating dma_segment_len in spi_dma_move_buffers() the following calculation is made:

https://github.com/zephyrproject-rtos/zephyr/blob/4256cd41df6c60f1832fd2deb14edc30ac7debab/drivers/spi/spi_ll_stm32.c#L208

For 8 bit SPI words this is inadvertently correct because "len" will be in terms of number of bytes and the "data->dma_rx.dma_cfg.dest_data_size" will always be equal to 1. Therefore "number_of_bytes/1" will give the correct dma_segment_len.

However for 16 bit SPI words "len" will now be given in half words rather than bytes and the "data->dma_rx.dma_cfg.dest_data_size" should be 2.  Therefore "number_of_half_words/2" will be 1/4 of the actual dma_segment_len. I think this should be changed to:

```
 dma_segment_len = len * data->dma_rx.dma_cfg.dest_data_size; 
 ...
 dma_segment_len = len * data->dma_tx.dma_cfg.source_data_size;
```

This bug has been previously noticed in this discussion #45348.

**To Reproduce**
1. kconfig:
CONFIG_SPI=y
CONFIG_SPI_STM32_DMA=y

2. dts:
&spi {
    dmas = <&dma1 5 1 0x32C40>,  // half word
    <&dma1 4 1 0x32C80>;
    dma-names = "tx", "rx";
}

3. SPI device:
SPI device with SPI_WORD_SET(16)

**Expected behavior**
16 bit SPI DMA transfers should work.

**Impact**
Annoyance, I must make my device's SPI word size 8 bits for DMA to work. 

**Environment (please complete the following information):**
Zephyr version https://github.com/zephyrproject-rtos/zephyr/commit/4256cd41df6c60f1832fd2deb14edc30ac7debab

### PR DESCRIPTION
**Pull request by rmsaad**
Calculates the correct dma segment length for STM32 16 bit spi dma transfers. Fixes zephyrproject-rtos#52563

Signed-off-by: Rami Saad <rami.saad@morgansolar.com>

### CODE REVIEW

**Comment by GeorgeCGV** [comment]
Hello, thank you for the contribution.

I don't think that proposed implementation is correct. The length seems to be controlled by the SPI context:
https://github.com/zephyrproject-rtos/zephyr/blob/ba7d730e9bd9250679904fd7d7724738e7c3603c/drivers/spi/spi_ll_stm32.c#L753-L759
the context sets the length within prior call to
https://github.com/zephyrproject-rtos/zephyr/blob/ba7d730e9bd9250679904fd7d7724738e7c3603c/drivers/spi/spi_ll_stm32.c#L732

The `spi_context_buffers_setup` has `dfs` parameter that stands for "Data Frame Size". The majority of SPI drivers don't use it. However, ESP32, Synopsys and Xilinix do. 

I suggest to investigate that approach.

**Comment by rmsaad** [comment]
> Hello, thank you for the contribution.
> 
> I don't think that proposed implementation is correct. The length seems to be controlled by the SPI context:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/ba7d730e9bd9250679904fd7d7724738e7c3603c/drivers/spi/spi_ll_stm32.c#L753-L759
> 
> 
> the context sets the length within prior call to
> https://github.com/zephyrproject-rtos/zephyr/blob/ba7d730e9bd9250679904fd7d7724738e7c3603c/drivers/spi/spi_ll_stm32.c#L732
> 
> The `spi_context_buffers_setup` has `dfs` parameter that stands for "Data Frame Size". The majority of SPI drivers don't use it (_i.e., STM32 implementation passes `1`_). However, ESP32, Synopsys and Xilinix do.
> 
> I suggest to investigate that approach.

I tried changing the dfs parameter from "1" to "2" and that halved the spi context's tx_len and rx_len variables. This is the opposite of what is needed because dma_len needs tx_len and rx_len in terms of bytes.

**Comment by rmsaad** [comment]
@GeorgeCGV, 

When utilizing spi_context_buffers_setup(), the spi context will divide the the spi_bufs "len" by "dfs" and set "tx_len" and "rx_len" to those values.

https://github.com/zephyrproject-rtos/zephyr/blob/4256cd41df6c60f1832fd2deb14edc30ac7debab/drivers/spi/spi_context.h#L272-L273

 For example, That means if the length of each spi_buf (rx and tx) is say 200 (half words), then passing a "dfs" of "2" will result in rx_len == 100 and tx_len == 100. However, when we set the block size (ie. dma_segment_len) for the DMA transfer it should equal 400. I don't think there is a intuitive way to achieve this by using the spi context.

**Comment by GeorgeCGV** [comment]
@rmsaad, I see. I mentioned `dfs` as it means "data frame size" and is being used in the SPI context to compute the length. That is more or less related to the problem. Therefore, it would be a good idea to investigate why it is there and if it is able to solve the problem.

>  if the length of each spi_buf (rx and tx) is say 200 (half words)

btw, `spi_buf` length is expected to be in bytes:
https://github.com/zephyrproject-rtos/zephyr/blob/997f5edf1a43252ffa644f471dddbfe7f9d99e47/include/zephyr/drivers/spi.h#L432-L434

**Comment by rmsaad** [comment]
@GeorgeCGV,

> btw, `spi_buf` length is expected to be in bytes:

I feel that spi_buf length documentation is very vague. The only in tree driver with "SPI_WORD_SET(16)" is setting the spi_buf length in 16 bit words:

https://github.com/zephyrproject-rtos/zephyr/blob/bba9fc98535567008728143cc3b8c029670d84c6/drivers/wifi/eswifi/eswifi_bus_spi.c#L91

Also if you look at the non-DMA code, the dfs is changed to "2" if operation is 16 bit even though it was setup earlier as 1. Therefore, the buffer would be pointing to the next address on a 16 bit offset from the previous, this takes account for spi_buf length being given in 16 bit words.   

https://github.com/zephyrproject-rtos/zephyr/blob/ba7d730e9bd9250679904fd7d7724738e7c3603c/drivers/spi/spi_ll_stm32.c#L282-L290

**Comment by GeorgeCGV** [comment]
The #4943 added 16-bit support, but left one comment untouched:
https://github.com/zephyrproject-rtos/zephyr/blob/1aebcec02ff9b901cab8056bc6168266033efe0a/drivers/spi/spi_ll_stm32.c#L349-L351

It could be great to remove it.

> I feel that spi_buf length documentation is very vague.

Indeed. Clarification is required. However, that is outside of this PR.

If `dma_cfg.dest_data_size` is `1` for 8-bit and `2` for 16-bit while the `len` is the number of frames (_1 for 8-bit, 1 for 16-bit_), then I don't see a problem in proposed fix. Nonetheless, I will wait for @FRASTM feedback.

**Comment by FRASTM** [comment]
> The #4943 added 16-bit support, but left one comment untouched:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/1aebcec02ff9b901cab8056bc6168266033efe0a/drivers/spi/spi_ll_stm32.c#L349-L351
> 
> It could be great to remove it.
> 
> > I feel that spi_buf length documentation is very vague.
> 
> Indeed. Clarification is required. However, that is outside of this PR.
> 
> If `dma_cfg.dest_data_size` is `1` for 8-bit and `2` for 16-bit while the `len` is the number of frames (_1 for 8-bit, 1 for 16-bit_), then I don't see a problem in proposed fix. Nonetheless, I will wait for @FRASTM feedback.

If the question is about to remove the `TODO:`  on  spi_stm32_shift_frames() function, I agree, this can be done, even if not really the purpose of that PR.

**Comment by rmsaad** [comment]
@FRASTM, do you see any inconsistencies in my reasoning for this PR?

**Comment by FRASTM** [comment]
> @FRASTM, do you see any inconsistencies in my reasoning for this PR?

This change make sense, to have dma_segment_len  twice the frame when data_size is 2 bytes long.
When the dma channel is configured, the DMA_InitStruct.NbData retrieves the `config->head_block->block_size / config->source_data_size; `   --> dma_segment_len
However I did not see any example demonstrating this, where the config of the dma channel has (peripheral or memory) data_size is _half-word_. That could be fine to have such a usecase.

**Comment by rmsaad** [comment]
@FRASTM,

I'm sorry, I am not quite following. Can you explain further on additional changes you want made?

**Comment by FRASTM** [comment]
> @FRASTM,
> 
> I'm sorry, I am not quite following. Can you explain further on additional changes you want made?

The change LGTM, I was just wondering if you add an example with stm32  spi transferring on 16bits

**Review by FRASTM — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**
